### PR TITLE
Constructor implementation

### PR DESCRIPTION
**Pull request by fredrikekre**

### CODE REVIEW

**Comment by fredrikekre** [comment]
This passes tests now, will clean up.

Now when functions return a `Tensor` we don't give the type `T`, it is determined by the data. That is quite nice, since we dont have to worry about what type the result will be.

Although I also added these:

``` jl
(::Type{Tensor{order, dim, T1, M}}){order, dim, T1 <: Real, T2 <: Real, M}(t::NTuple{M, T2}) = Tensor{order, dim}(SVector{M,T1}(t))
```

so technically the old methods would work anyway...

Whats your opinion?

**Comment by fredrikekre** [comment]
Will do benchmarking now

File: src/ContMechTensors.jl [hunk 1]
```diff
@@ -315,7 +352,7 @@ for TensorType in (SymmetricTensor, Tensor)
             exp = tensor_create(get_type(Tt),f)
             return quote
                 $(Expr(:meta, :inline))
-                $($TensorType){order, dim, T, $N}($exp)
```
Comment (KristofferC) [inline src/ContMechTensors.jl:318]: Can the `N` in this function a few lines up be removed? It doesn't seem to be used. Same for other functions.

Comment (fredrikekre) [reply src/ContMechTensors.jl:318]: Yea I am cleaning up. Also all the `Tv` we calculate can be removed with this method, which I like

Comment (KristofferC) [reply src/ContMechTensors.jl:318]: Yeah, thats nice.

**Comment by fredrikekre** [comment]
https://gist.github.com/fredrikekre/d077647e8e19b8521523a8ce678fd61b

Hmmm

**Comment by fredrikekre** [comment]
https://gist.github.com/fredrikekre/468bb9b3f805495c6478ce9d37371d24

:sunglasses:

**Comment by fredrikekre** [comment]
ALL of the sym stuff got so much faster hehe

**Comment by KristofferC** [comment]
So much improvement on a common operation like dot product between two symmetric second order tensors is really nice.

**Comment by KristofferC** [comment]
Ready to go?

**Comment by fredrikekre** [comment]
No, will fix some more stuff

**Comment by KristofferC** [comment]
The code on master looks really bad for symmetric tensors for some reason, which explains the speedups here. Why didnt this show in the benchmarks?

``` jl
julia> @code_llvm a ⋅ a
; Function Attrs: uwtable
define void @julia_dot_72004(%Tensor* noalias sret, %SymmetricTensor*, %SymmetricTensor*) #0 {
top:
  %3 = alloca %SMatrix, align 8
  %4 = alloca %Tensor, align 8
  %5 = getelementptr inbounds %SymmetricTensor, %SymmetricTensor* %1, i64 0, i32 0, i32 0, i64 0
  %6 = load double, double* %5, align 1
  %7 = getelementptr inbounds %SymmetricTensor, %SymmetricTensor* %1, i64 0, i32 0, i32 0, i64 1
.
.
  %18 = fmul double %8, %14
  %19 = fadd double %17, %18
  %20 = fmul double %8, %12
.
.
  %28 = getelementptr inbounds %SMatrix, %SMatrix* %3, i64 0, i32 0, i64 0
  store double %19, double* %28, align 8
  %29 = getelementptr inbounds %SMatrix, %SMatrix* %3, i64 0, i32 0, i64 1
.
.
  call void @julia_Type_72006(%Tensor* noalias nonnull sret %4, %jl_value_t* inttoptr (i64 2240490384 to %jl_value_t*), %SMatrix* nonnull %3) #1
  %32 = bitcast %Tensor* %0 to i8*
  %33 = bitcast %Tensor* %4 to i8*
  call void @llvm.memcpy.p0i8.p0i8.i64(i8* %32, i8* %33, i64 32, i32 8, i1 false)
  ret void   
```

**Comment by fredrikekre** [comment]
Some change in StaticArrays between then and now?

**Comment by KristofferC** [comment]
Ah, maybe! The dangers of relying on other packages.

**Comment by fredrikekre** [comment]
PR vs current master:
https://gist.github.com/fredrikekre/5a579130722ff9b007b67bde6e8a17ca

PR vs before StaticArray
https://gist.github.com/fredrikekre/0f0f6bc8f30a98f4de5311b7f8d5b834

File: src/ContMechTensors.jl [hunk 2]
```diff
@@ -75,56 +67,31 @@ function tomatrix{dim}(t::Tensor{2, dim})
     return SMatrix{N, N}(tovector(t))
 end
 
-function n_independent_components(dim, issym)
```
Comment (fredrikekre) [inline src/ContMechTensors.jl:78]: Used once.

File: src/ContMechTensors.jl [hunk 3]
```diff
@@ -135,10 +102,28 @@ Base.similar(t::AbstractTensor) = t
...
+        @inline (::Type{$TensorType{order, dim}}){order, dim, T <: Real, M}(t::NTuple{M, T}) = $TensorType{order, dim}(SVector{M,T}(t))
+        @inline (::Type{$TensorType{order, dim, T1}}){order, dim, T1 <: Real, T2 <: Real, M}(t::NTuple{M, T2}) = $TensorType{order, dim}(SVector{M,T1}(t))
+        @inline (::Type{$TensorType{order, dim, T1, M}}){order, dim, T1 <: Real, T2 <: Real, M}(t::NTuple{M, T2}) = $TensorType{order, dim}(SVector{M,T1}(t))
+
+        @inline (::Type{$TensorType{order, dim}}){order, dim, T <: Real, M}(t::SVector{M, T}) = throw(ArgumentError("error"))
+    end
+    for order in (2,4), dim in (1,2,3)
+        M = n_components(TensorType{order,dim})
+        @eval @inline (::Type{$TensorType{$order, $dim}}){T <: Real}(t::SVector{$M, T}) = $TensorType{$order, $dim, T, $M}(t)
+        @eval @inline (::Type{$TensorType{$order, $dim}}){T <: Real, Q}(t::SMatrix{Q, Q, T, $M}) = $TensorType{$order, $dim, T, $M}(t.data)
```
Comment (fredrikekre) [inline src/ContMechTensors.jl:125]: this might add some non-used constructors for `SMatrix` but w/e

File: src/symmetric_ops.jl [hunk 4]
```diff
@@ -6,13 +6,13 @@
 # Double contraction #
 ######################
 
-@gen_code function dcontract{dim, T1, T2}(S1::SymmetricTensor{2, dim, T1}, S2::SymmetricTensor{2, dim, T2})
+@gen_code function dcontract{dim, T1, T2, M}(S1::SymmetricTensor{2, dim, T1, M}, S2::SymmetricTensor{2, dim, T2, M})
```
Comment (fredrikekre) [inline src/symmetric_ops.jl:9]: adding this `M` was all that was needed instead of `n_independent_components` some lines below heh

**Review by fredrikekre — COMMENTED** [review]
Lots of whitespace changes, so use `?w=1`

**Comment by fredrikekre** [comment]
> The dangers of relying on other packages.

We just have to run some benchmarking now and then I guess.

File: src/ContMechTensors.jl [hunk 5]
```diff
@@ -105,13 +105,19 @@ Base.fill(t::AbstractTensor, v::Number) = one(typeof(t)) * v
 #########################
 # Internal constructors #
 #########################
+function _constructor_check{order, dim}(Tt::Union{Type{Tensor{order, dim}}, Type{SymmetricTensor{order, dim}}})
```
Comment (fredrikekre) [inline src/ContMechTensors.jl:108]: can this be `@pure`?

Comment (KristofferC) [reply src/ContMechTensors.jl:108]: No. Basically, if the value only depends on the type then it can be pure.

File: src/ContMechTensors.jl (hunk 3 above)
Comment (KristofferC) [reply src/ContMechTensors.jl:125]: Doesnt matter.

**Comment by KristofferC** [comment]
https://gist.github.com/KristofferC/e0bd470f1f9accba53cb6b88ffca4038

This PR on newest StaticArrays vs this PR on the one we did previous benchmarks on.

File: src/ContMechTensors.jl [hunk 6]
```diff
@@ -283,11 +255,9 @@ end
 ##########################
 # Zero, one, rand, diagm #
```
Comment (KristofferC) [inline src/ContMechTensors.jl]: This comment is a bit wrong :P

File: src/ContMechTensors.jl [hunk 7]
```diff
@@ -250,20 +223,19 @@ for op in (:zero, :rand)
         @eval begin
             @inline function Base.$op{order, dim, T}(Tt::Type{$TensorType{order, dim, T}})
                 N = n_components($TensorType{order, dim})
-                return $TensorType{order, dim, T, N}($op(SVector{N, T}))
+                return $TensorType{order, dim}($op(SVector{N, T}))
```
Comment (KristofferC) [inline src/ContMechTensors.jl:191]: This is wrong now (element type):

``` jl
julia> a = rand(Vec{3, Float32})
3-element ContMechTensors.Tensor{1,3,Float64,3}:
 0.0519396
 0.944041 
 0.438569 
```

Need to keep the fact that the user explicitly asked for `T`. Should probably add as a test.

File: src/ContMechTensors.jl (hunk 5 above)
Comment (fredrikekre) [reply src/ContMechTensors.jl:108]: But it does?

**Comment by KristofferC** [comment]
This doesn't work:

``` jl
julia> Vec{3, Float32}(i->i+1)
------ ErrorException ------------------ Stacktrace (most recent call last)

 [1] — ContMechTensors.Tensor{1,3,Float32,3}(::Function) at ContMechTensors.jl:26

No precise constructor found. Length of input was 1 while length of StaticArrays.SVector{3,Float32} is 3.
```

Might need a special case for Vec.

File: src/ContMechTensors.jl (hunk 7 above)
Comment (fredrikekre) [reply src/ContMechTensors.jl:191]: I messed it up, will fix.

Comment (fredrikekre) [reply src/ContMechTensors.jl:191]: Should be tested btw

**Comment by fredrikekre** [comment]
Yes

File: src/ContMechTensors.jl (hunk 7 above)
Comment (fredrikekre) [reply src/ContMechTensors.jl:191]: We rand a SVector of T though? I think it is just for vec. Goes through the func def below.

**Comment by fredrikekre** [comment]
``` jl
julia> Vec{3, Float32}(i->i+1)
ERROR: No precise constructor found. Length of input was 1 while length of StaticArrays.SVector{3,Float32} is 3.
 in ContMechTensors.Tensor{1,3,Float32,3}(::Function) at /home/fredrik/Dropbox/Programming/ContMechTensors.jl/src/ContMechTensors.jl:26
```

Hmmm.... This actually didn't work. Since `Vec{3, Float32}` is a `Tensor{1, 3, Float32, 3}` it calls the type directly....

``` jl
julia> @which Vec{3, Float32}(i->i+1)
(::Type{ContMechTensors.Tensor{order,dim,T,M}}){order,dim,T<:Real,M}(data) at /home/fredrik/Dropbox/Programming/ContMechTensors.jl/src/ContMechTensors.jl:26
```

**Comment by fredrikekre** [comment]
Thought adding this would work, but it doesnt...

``` jl
@inline (::Type{Vec{dim, T}}){dim, T}(f::Function) = Tensor{1, dim, T}(f)
```

File: src/ContMechTensors.jl [hunk 8]
```diff
@@ -135,107 +102,72 @@ Base.similar(t::AbstractTensor) = t
...
+# General fallbacks
+@inline function (Tt::Union{Type{Tensor{order, dim, T}}, Type{SymmetricTensor{order, dim, T}}}){order, dim, T}(data)
+    t1 = get_base(Tt)(data)
+    return convert(Tt, t1)
 end
 
-function (Tt::Union{Type{Tensor{order, dim, T}}, Type{SymmetricTensor{order, dim, T}}}){order, dim, T}(f_or_data)
-    t1 = get_main_type(Tt){order, dim}(f_or_data)
-    return convert(get_main_type(Tt){order, dim, T}, t1)
+@inline function (Tt::Union{Type{Tensor{order, dim, T, M}}, Type{SymmetricTensor{order, dim, T, M}}}){order, dim, T, M}(data)
```
Comment (fredrikekre) [inline src/ContMechTensors.jl:142]: This will never be callable either.

**Comment by fredrikekre** [comment]
before StaticArray:

``` jl
julia> Vec{3, Float32}(i->i+1)
ERROR: MethodError: Cannot `convert` an object of type ##1#2 to an object of type Tuple{Float32,Float32,Float32}
This may have arisen from a call to the constructor Tuple{Float32,Float32,Float32}(...),
since type constructors fall back to convert methods.
 in ContMechTensors.Tensor{1,3,Float32,3}(::Function) at /home/fredrik/Dropbox/Programming/ContMechTensors.jl/src/ContMechTensors.jl:27
```

**Comment by fredrikekre** [comment]
PR vs current master:
https://gist.github.com/fredrikekre/5a579130722ff9b007b67bde6e8a17ca

PR vs before StaticArray
https://gist.github.com/fredrikekre/0f0f6bc8f30a98f4de5311b7f8d5b834

**Comment by KristofferC** [comment]
Gooooo.

**Comment by KristofferC** [comment]
Sweeet

**Pull request merged**